Add Python bindings

tetsuo-cpp opened the pull request:
Closes #20

@znewman01 @woodruffw @di

tetsuo-cpp commented on python/pyproject.toml:
```diff
@@ -0,0 +1,42 @@
+[project.optional-dependencies]
+dev = [
+  "betterproto[compiler]",
+  "grpcio-tools",
```
To get things working, I'm listing `grpcio-tools` as a dependency and using the bundled `protoc` instead of invoking the Docker container.

Before this gets merged in, I can fiddle with this to plug `betterproto` into the Docker container's `protoc`. Or we can avoid the Docker container altogether and use this `protoc` for every language.

tetsuo-cpp commented on python/sigstore_protobuf_specs/dev/sigstore/bundle/v1.py:
```diff
@@ -0,0 +1,63 @@
+
+import betterproto
+
+from .dev.sigstore.common import v1
```
Unfortunately these relative imports are busted. I'm trying to figure out whether this is a case of PEBCAK or whether there's a bug that needs fixing in `betterproto`.

tetsuo-cpp commented on Makefile:
```diff
@@ -23,6 +23,17 @@ go:
+	python3 -m venv env
+	./env/bin/python -m pip install --upgrade pip
+	./env/bin/python -m pip install -e ./python[dev]
+	./env/bin/python -m pip install grpcio-tools
```
At the moment, I'm installing `grpcio-tools` to invoke the Protobuf compiler.

Before we merge this, I can either fiddle with this to plug `betterproto` into the Docker image's compiler or we can remove the Docker container altogether and make everything use this one.

woodruffw replied:
I'm personally okay with that, although IMO the `grpcio-tools` dependency should go in the `[dev]` extra rather than being a standalone dependency here.

tetsuo-cpp replied:
Since we'll be moving to executing this within Docker, I think I'll just use the `protoc` in the image.

woodruffw replied:
Sounds good!

tetsuo-cpp commented on gen/pb-python/pyproject.toml:
```diff
@@ -0,0 +1,41 @@
+
+[project.optional-dependencies]
+dev = [
+  "betterproto[compiler]==2.0.0b5",
```
The developers of `betterproto` have been working on a v2.0.0 release for a few years now. We pretty much have to use a beta version as the current stable version seems to have bugs with relative imports and lacks support for `optional`.

woodruffw replied:
I'm okay with a beta version here, as long as we keep it pinned (which we are).

tetsuo-cpp commented:
Since the specs aren't finalised yet, I haven't added a release workflow to build wheels, publish to PyPI, etc. But that should be pretty trivial to do when we need it.

woodruffw commented:
> Since the specs aren't finalised yet, I haven't added a release workflow to build wheels, publish to PyPI, etc. But that should be pretty trivial to do in a follow up PR when we need it.

Makes sense! IMO we should also get a skeleton test suite in here before doing an initial release.

znewman01 commented on Makefile:
```diff
@@ -23,6 +23,17 @@ go:
 	@echo "Generating go protobuf files"
 	docker run --pull always --platform linux/amd64 -v ${PWD}:/defs ${PROTOC_IMAGE} -d protos -l go --go-module-prefix github.com/sigstore/protobuf-specs/gen/pb-go
 
+python: env/pyenv.cfg
```
Can you add `python` to the `all:` target?

tetsuo-cpp replied:
[ffa2192](https://github.com/sigstore/protobuf-specs/pull/25/commits/ffa2192f5cc476fff26ef0f26f5fec413c0b17a5)

znewman01 commented on gen/pb-python/pyproject.toml:
```diff
@@ -0,0 +1,41 @@
+
+[project]
+name = "sigstore-protobuf-specs"
+version = "0.0.1"
```
Note to self (see https://github.com/sigstore/protobuf-specs/issues/22): at some point we're going to want one canonical source of truth for the version number. Or at least a "bump version script" that will update this.

(You don't need to worry for this PR)

woodruffw replied:
+1 -- in the past I've done this with a top-level `VERSION` file, although the downside to that is that it doesn't play nicely with `pyproject.toml`.

woodruffw replied:
(We could also keep it synced automatically with a GitHub Action).

tetsuo-cpp replied:
Created #26 to track this.

znewman01 commented on Makefile:
```diff
@@ -23,6 +23,17 @@ go:
+	@echo "Building virtual environment for python protobuf"
+	python3 -m venv env
+	./env/bin/python -m pip install --upgrade pip
+	./env/bin/python -m pip install -e ./gen/pb-python[dev]
```
These components don't appear to pin the build tool versions.

znewman01 replied:
> (The `grpcio-tools` dep should definitely be pinned, and we should do that while adding it to `pyproject.toml` per [#25 (comment)](https://github.com/sigstore/protobuf-specs/pull/25#discussion_r1028272826))

That's the one I really care about.

znewman01 commented on Makefile:
```diff
@@ -23,6 +23,17 @@ go:
+	@echo "Generating python protobuf files"
+	. env/bin/activate && cd ./gen/pb-python/sigstore_protobuf_specs && python -m grpc_tools.protoc -I ../../../protos/ --python_betterproto_out=. ../../../protos/*
+
+env/pyenv.cfg:
```
Though venvs are definitely a *better* solution for Python than Docker, they're a less consistent one. Ideally, you would be able to clone this repo on a machine without Python installed and still update it with a simple `make all`. Is is possible to move these steps *inside* of a Docker container? (You can even keep them exposed as Make targets to run them directly if you'd like)

znewman01 reviewed (commented):
TYVM, this is looking great.

Couple of nits, mostly having to do with "can you do things in strictly worse ways because we have to share this repo?"

tetsuo-cpp commented:
Ok! That was a bit more complicated than expected. I've modified the Python build to use `protoc` within the Docker container and removed the use of virtual environments.

I'll leave a few comments to explain why I've done certain things.

tetsuo-cpp commented on Dockerfile:
```diff
@@ -1,3 +1,12 @@
+RUN set -ex && \
+    apt-get update && \
+    apt-get install -y --no-install-recommends \
+    python3-pip
```
Because we're executing the plugin within Docker, I need to install the `betterproto` in the image. For that, I need `pip`.

Therefore, I've had to begin building a new image on top of `namely/protoc-all` since the base image no longer has everything we need. This is covered by the Makefile.

tetsuo-cpp commented on Makefile:
```diff
@@ -13,19 +13,38 @@
+
+python: docker-image
+	@echo "Generating python protobuf files"
+# we need to manually fix the PYTHONPATH due to: https://github.com/namely/docker-protoc/pull/356
```
Ran into an ugly issue which I've got a fix for here: https://github.com/namely/docker-protoc/pull/356

Basically, the buggy PYTHONPATH causes the interpreter to look in the current directory for modules. And since we have a module called `io` at the top level of the `sigstore_protobuf_specs`, the interpreter actually imports that instead the standard library `io` on startup.

znewman01 reviewed (approved):
This LGTM, thanks so much!
